lib/spack/spack/repo.py

```diff
@@ -448,7 +451,7 @@ def __init__(self, *repos, **kwargs):
         for repo in repos:
             try:
                 if isinstance(repo, string_types):
-                    repo = Repo(repo, self.super_namespace)
```

Comment:
@alalazo so this all looks reasonable in that it should consider all package objects as part of the `spack.pkg` module, but the code before this also looks like it should do that. This is another way of saying that as I noted in https://github.com/spack/spack/issues/11159#issuecomment-484228264, I don't think the core problem has been identified.

That being said, if you added a unit test which checked for directives being applied to packages in alternative namespaces, and it passed here (and it fails for the current logic), then I think that would be sufficient reason for merging this. Let me know if you want to discuss how to do that.

Reply:
Summarizing a few things to check if my understanding of the code is correct. It seems to me that the `namespace` argument we can currently pass to `Repo` and `RepoPath` actually refers to the super-namespace i.e. usually `spack.pkg`, while the "real" namespace of the package is read from `repos.yaml` and usually coincides with the enclosing folder. 

To be less abstract the builtin repository is `spack.pkg.builtin` where `spack.pkg` is the super-namespace and `builtin` is the namespace. The intent of this PR is to prevent clients from being able to set the super-namespace at all, as that will create repositories with invalid packages due to the check in `directives.py`.

> That being said, if you added a unit test which checked for directives being applied to packages in alternative namespaces, and it passed here (and it fails for the current logic), then I think that would be sufficient reason for merging this. Let me know if you want to discuss how to do that.

I'm all in to add tests to this, but I think you're requiring the other fix I was mentioning in #11159 i.e. relaxing the constraint in `directives.py` and permit other super-namespaces. Am I getting this wrong?

Reply:
> I'm all in to add tests to this, but I think you're requiring the other fix I was mentioning in #11159 i.e. relaxing the constraint in directives.py and permit other super-namespaces.

I don't think other super namespaces should be allowed (or should have to be supported). Maybe what I was getting confused by is that there are many changes and perhaps the only necessary changes are very small:

Now that I read through all of this again, it looks like the primary functional change is to remove the `namespace` argument from the `Repo` initializer; this prevents users from accidentally assigning a different super-namespace (which as you note in #11159 is not supported). So if that's true, then no test is needed because we are just removing functionality.

Also if that's true: please put the necessary change in a separate commit, or clarify in the PR description that this is the required change to address the issue.

Reply:
> Also if that's true: please put the necessary change in a separate commit, or clarify in the PR description that this is the required change to address the issue.

Sorry, the commits are well factored but it's true that the description was not useful at all. I changed that to reflect the changes better. The only necessary change to fix #11159 is 429b811 while the remaining three commits seemed to me easy and necessary changes to ease later maintenance. Let me know if something else is needed by my side.